### PR TITLE
RFC: Global Conditions

### PR DESCRIPTION
**Pull request by adpaco-aws**
### Description of changes: 

This RFC proposes a new section in Kani's output to summarize the status of properties that depend on other properties, referred to as "Global Conditions". Originally, I wanted this change to be part of #2299, but then I realized a small RFC might be appropriate.

### Resolved issues:

N/A

### Testing:

* How is this change tested? N/A

* Is this a refactor change? N/A

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by celinval — COMMENTED** [review]
It makes sense.

**Comment by zhassan-aws** [comment]
Link to rendered version: https://github.com/adpaco-aws/rmc/blob/update-other-rfcs/rfc/src/rfcs/0007-global-conditions.md

File: rfc/src/rfcs/0007-global-conditions.md
Comment (zhassan-aws) [inline rfc/src/rfcs/0007-global-conditions.md:17]: Is this a standard definition of hyperproperties? The definition I'm aware of is that while properties refer to traces, hyperproperties refer to sets of traces:
https://en.wikipedia.org/wiki/Hyperproperty
The definition used here doesn't seem to match with that.

**Review by karkhaz — COMMENTED** [review]
It seems to me that there are two commonalities in @adpaco-aws's examples of `should_panic` and `fail_uncoverable`:

* Their statuses are derived entirely from the status of other properties, and not directly a result from the SAT solver
* They are of the form "{panic,coverage} did not happen along any path" rather than "{overflow} happened along some path".

Would you expect these to be true for other global conditions that might be added in the future? I think the first bullet point is how you're defining hyperproperty/global condition, but I think the second bullet point is more relevant to users and might be a good reason to separate hyperproperties out from regular ones.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (celinval) [reply rfc/src/rfcs/0007-global-conditions.md:78]: Will unwinding assertions fall into the global conditions category?

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md]: Done.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md]: Done, see [this comment](https://github.com/model-checking/kani/pull/2516#discussion_r1229770464).

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md]: Sorry about that, please see [this comment](https://github.com/model-checking/kani/pull/2516#discussion_r1229770464).

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:30]: The RFC now explicitly mentions the terse output now. My proposal is to only include the last change (i.e., the one that changes overall result).

**Comment by adpaco-aws** [comment]
> I cannot think of other "Global Conditions" that may be introduced in the future. Do you have any in mind?

 Yes, the check for expectations on individual properties mentioned [here](https://model-checking.github.io/kani/rfc/rfcs/0005-should-panic-attr.html#alternative-2-name) could be implemented as a global condition. Another global condition that I think could be interesting for users would check that specific properties (e.g., user-defined assertions) are not unreachable. This can already be manually done by adding `cover` statements, but that isn't always an option (because the user may not want to add them) and users may not realize they are unreachable if they appear in deep places in the call graph.

But one of the cool things about global conditions is that we could also define them for existing properties. For example, we could have a `memory safety` condition that summarizes the status memory-safety checks (and hide them from `RESULTS` if all went well). Or do the same for unwinding assertions, as @celinval mentioned. If we did that, this section could become the main component of Kani's output.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md]: I've clarified this with a footnote.

**Comment by celinval** [comment]
> > I cannot think of other "Global Conditions" that may be introduced in the future. Do you have any in mind?
> 
> Yes, the check for expectations on individual properties mentioned [here](https://model-checking.github.io/kani/rfc/rfcs/0005-should-panic-attr.html#alternative-2-name) could be implemented as a global condition. Another global condition that I think could be interesting for users would check that specific properties (e.g., user-defined assertions) are not unreachable. This can already be manually done by adding `cover` statements, but that isn't always an option (because the user may not want to add them) and users may not realize they are unreachable if they appear in deep places in the call graph.
> 
> But one of the cool things about global conditions is that we could also define them for existing properties. For example, we could have a `memory safety` condition that summarizes the status memory-safety checks (and hide them from `RESULTS` if all went well). Or do the same for unwinding assertions, as @celinval mentioned. If we did that, this section could become the main component of Kani's output.

Please do replace the memory safety and arithmetic checks by global conditions! :) It would result in a much cleaner output.

One thing I'm missing in this RFC is an explanation on how the user could easily find the root cause of the failing global condition. For example, for the `fail_uncovered`, do we expect users to go inspect all the covers or can we print the uncovered ones the same way we print failed checks at the end?

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md]: I've changed this to "may". My take on this is that global condition condense information that involves several checks, so they shouldn't be displayed as regular ones. But I don't have any evidence, it's still a subjective statement.

**Comment by adpaco-aws** [comment]
> They are of the form "{panic,coverage} did not happen along any path" rather than "{overflow} happened along some path".

Actually, the `should_panic` condition expresses "there exists at least one path with an assertion-related failure". I'm defining global conditions as in the first item because that's what I'm trying to generalize here, and I'd expect new ones to follow this rule but I can't guarantee it. But that's OK, we can always go back to this RFC and update it 😄

**Comment by adpaco-aws** [comment]
> One thing I'm missing in this RFC is an explanation on how the user could easily find the root cause of the failing global condition. For example, for the `fail_uncovered`, do we expect users to go inspect all the covers or can we print the uncovered ones the same way we print failed checks at the end?

That's a great point, but I'm concerned about the output getting longer and less comprehensible if we include that change in this RFC. My preference here would be to keep it simple for now, but I'm happy to talk about it.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:79]: Sorry, missed your comments above. My answers:
 1. No, this wouldn't change how `should_panic` behaves. It would only change its UI.
 2. Yes, a global condition can be used to represent the result of all unwinding conditions.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:81]: It's actually less than 200 LoC now that I've checked. I'll remove this sentence.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:30]: Yeah, then I think it's better to not distinguish here and just say that the "global conditions" section will be printed with all failing global conditions for the terse output. That'd be the only difference at this point.

**Comment by celinval** [comment]
> > One thing I'm missing in this RFC is an explanation on how the user could easily find the root cause of the failing global condition. For example, for the `fail_uncovered`, do we expect users to go inspect all the covers or can we print the uncovered ones the same way we print failed checks at the end?
> 
> That's a great point, but I'm concerned about the output getting longer and less comprehensible if we include that change in this RFC. My preference here would be to keep it simple for now, but I'm happy to talk about it.

I agree that our output is pretty long and it needs refactoring ASAP.

However, I think that printing the uncovered statements at the end will alleviate the problem of our super verbose output.  At least the information that the user really needs is readily available at the end of the report, just like failed properties.

**Comment by adpaco-aws** [comment]
> However, I think that printing the uncovered statements at the end will alleviate the problem of our super verbose output. At least the information that the user really needs is readily available at the end of the report, just like failed properties.

Makes sense. I've been extending the current implementation (in #2532) to use the function that reports failed checks for standard verification jobs, but it needs improvements. For example, for a simple harness which contains an unreachable `cover` statement we get this information:

```
RESULTS:
Check 1: harness_uncoverable.cover.1
         - Status: UNREACHABLE
         - Description: "cover location"
         - Location: tests/ui/fail-uncoverable/uncoverable/test.rs:13:9 in function harness_uncoverable

GLOBAL CONDITIONS:
 - fail_uncoverable: FAILURE (encountered one or more cover statements which were not satisfied)
Failed Checks: cover location

SUMMARY:
 ** 0 of 0 failed

 ** 0 of 1 cover properties satisfied (1 unreachable)


VERIFICATION:- FAILED (one or more global conditions failed)
Verification Time: 0.02321334s
```

Note that it only says "Failed Checks: cover location", but it doesn't attach a location despite showing it for the property because the trace doesn't refer to this location, causing us to return earlier. We should make some changes here, also because the "Failed Checks: " gets printed for each item (the results below are for `tests/ui/should-panic-attribute/unexpected-failures/test.rs`) which is causing the output to be cluttered and a little confusing:

```
SUMMARY:
 ** 3 of 18 failed
Failed Checks: panicked on the `1` arm!
 File: "/home/ubuntu/kani/tests/ui/should-panic-attribute/unexpected-failures/test.rs", line 18, in check
Failed Checks: panicked on the `0` arm!
 File: "/home/ubuntu/kani/tests/ui/should-panic-attribute/unexpected-failures/test.rs", line 17, in check
Failed Checks: shift distance too large
```
 In any case, I agree with your point - I'll change the RFC with a new proposal to report these failed checks.

**Review by celinval — APPROVED** [review]

**Comment by celinval** [comment]
Since this is an RFC, please wait for a second approval before merging.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (qinheping) [inline rfc/src/rfcs/0007-global-conditions.md:66]: How will users specify and enable global conditions? Can you give an example here?

Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:66]: Added a clarification regarding this. As mentioned, global conditions won't force a single mechanism to be enabled. This will rather depend on the specific feature (attributes, CLI options, or combinations of both).

**Review by karkhaz — APPROVED** [review]
Awesome. Thanks for working on this!!

**Review by jaisnan — APPROVED** [review]
Thanks :) !

File: rfc/src/rfcs/0007-global-conditions.md
Comment (remi-delmas-3000) [reply rfc/src/rfcs/0007-global-conditions.md:12]: So are global conditions solely defined as conjunctions of assert/cover statement or are there other ways of defining them ?

File: rfc/src/rfcs/0007-global-conditions.md
Comment (remi-delmas-3000) [inline rfc/src/rfcs/0007-global-conditions.md:58]: Is it possible to list the basic conditions that failed to explain why a global condition failed ?

File: rfc/src/rfcs/0007-global-conditions.md
Comment (remi-delmas-3000) [inline rfc/src/rfcs/0007-global-conditions.md:24]: is "hyperproperties" meant as in "a property that relates two or more executions of the same program?"

File: rfc/src/rfcs/0007-global-conditions.md
Comment (remi-delmas-3000) [inline rfc/src/rfcs/0007-global-conditions.md:55]: Are global conditions only enabled using special flags or can the user also define his own groups of dependent conditions ?

File: rfc/src/rfcs/0007-global-conditions.md
Comment (remi-delmas-3000) [inline rfc/src/rfcs/0007-global-conditions.md:83]: the documentation of `should_panic` does not really give a precise meaning to the check, i.e. it does not say if at least one execution must trigger a panic or if all executions must trigger a panic. How can it be modelled in terms of dependent conditions  and statuses ?

File: rfc/src/rfcs/0007-global-conditions.md
Comment (remi-delmas-3000) [inline rfc/src/rfcs/0007-global-conditions.md:90]: What is a regular check ? kani::assert ? kani::cover ?

File: rfc/src/rfcs/0007-global-conditions.md
Comment (remi-delmas-3000) [inline rfc/src/rfcs/0007-global-conditions.md:95]: is it possible to give a more explicit and rigorous definition of global conditions ? 

e.g. "a global condition is specified by a set of assert or cover statements and an aggregation function that computes a status for the global conditions from the statuses of the basic conditions."

File: rfc/src/rfcs/0007-global-conditions.md
Comment (remi-delmas-3000) [inline rfc/src/rfcs/0007-global-conditions.md:101]: so the status of a global condition could differ when the switch `--fail-uncoverable` is activated vs not activated ?

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:83]: I don't think we do that for any of the checks we currently produce. Regardless, I think it's well defined in [here](https://model-checking.github.io/kani/reference/attributes.html#kanishould_panic), and if you think otherwise I'm happy to do any editions you consider appropriate.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:90]: Check is the simpler term we use to refer to properties. Running Kani on any will output something like this:
```
RESULTS:
Check 1: example.assertion.1
         - Status: <status>
         - Description: <description>
         - Location: <location>
[...]
```

So this refers to those checks/properties.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:95]: Maybe? But it's definitely not that.

I'm actually thinking about growing the scope of global conditions to consider things other than other properties. For example, a timeout feature could be implemented a global condition. However, the current definition of global condition does not consider resources like memory or time, only verification properties.

File: rfc/src/rfcs/0007-global-conditions.md
Comment (adpaco-aws) [reply rfc/src/rfcs/0007-global-conditions.md:101]: No. The status computed for the condition itself won't differ. But activating the switch will simply cause the status of the global condition to affect the overall verification result.

**Pull request merged**
